testing/integration/upgrade_fleet_test.go

```diff
@@ -341,6 +341,135 @@ func testFleetAirGappedUpgrade(t *testing.T, stack *define.Info, unprivileged bo
+	require.NoError(t, err, "failed getting Fleet Server URL")
+
+	t.Logf("Installing Elastic Agent (privileged)...")
+	var nonInteractiveFlag bool
+	if upgradetest.Version_8_2_0.Less(*startParsedVersion) {
+		nonInteractiveFlag = true
+	}
+	installOpts := atesting.InstallOpts{
+		NonInteractive: nonInteractiveFlag,
+		Force:          true,
+		EnrollOpts: atesting.EnrollOpts{
+			URL:             fleetServerURL,
+			EnrollmentToken: enrollmentToken.APIKey,
+		},
+		Privileged: true,
+	}
+	output, err := startFixture.Install(ctx, &installOpts)
+	require.NoError(t, err, "failed to install start agent [output: %s]", string(output))
+
+	t.Log("Waiting for Agent to be correct version and healthy...")
+	err = upgradetest.WaitHealthyAndVersion(ctx, startFixture, startVersionInfo.Binary, 2*time.Minute, 10*time.Second, t)
+	require.NoError(t, err)
+
+	agentID, err := startFixture.AgentID(ctx)
+	require.NoError(t, err)
+	t.Logf("Agent ID: %q", agentID)
+
+	t.Logf("Upgrading from version \"%s-%s\" to version \"%s-%s\"...",
+		startParsedVersion, startVersionInfo.Binary.Commit,
+		endVersionInfo.Binary.String(), endVersionInfo.Binary.Commit)
+	err = fleettools.UpgradeAgent(ctx, kibClient, agentID, endVersionInfo.Binary.String(), true)
+	require.NoError(t, err)
+
+	t.Log("Waiting for upgrade details to show up in Fleet")
+	var agent kibana.GetAgentResponse
+	require.Eventuallyf(t, func() bool {
+		agent, err = kibClient.GetAgent(ctx, kibana.GetAgentRequest{ID: agentID})
+		return err == nil && agent.UpgradeDetails != nil
+	},
+		5*time.Minute, time.Second,
+		"last error: %v. agent.UpgradeDetails: %s",
+		err, agentUpgradeDetailsString(agent))
+
+	t.Log("Waiting for agent to have empty upgrade details")
+	require.Eventuallyf(t, func() bool {
+		agent, err = kibClient.GetAgent(ctx, kibana.GetAgentRequest{ID: agentID})
+		return err == nil && agent.UpgradeDetails == nil
+	},
+		20*time.Minute, time.Second,
+		"last error: %v. agent.UpgradeDetails: %s",
+		err, agentUpgradeDetailsString(agent))
+
+	t.Logf("Duplicate upgrade from version \"%s-%s\" to version \"%s-%s\"...",
+		startParsedVersion, startVersionInfo.Binary.Commit,
+		endVersionInfo.Binary.String(), endVersionInfo.Binary.Commit)
+	err = fleettools.UpgradeAgent(ctx, kibClient, agentID, endVersionInfo.Binary.String(), true)
+	require.NoError(t, err)
+
+	t.Log("Waiting for upgrade details to show up in Fleet")
+	require.Eventuallyf(t, func() bool {
+		agent, err = kibClient.GetAgent(ctx, kibana.GetAgentRequest{ID: agentID})
+		return err == nil && agent.UpgradeDetails != nil
+	},
+		5*time.Minute, time.Second,
+		"last error: %v. agent.UpgradeDetails: %s",
+		err, agentUpgradeDetailsString(agent))
+
+	t.Log("Waiting for agent to have empty upgrade details")
+	require.Eventuallyf(t, func() bool {
+		agent, err = kibClient.GetAgent(ctx, kibana.GetAgentRequest{ID: agentID})
+		return err == nil && agent.UpgradeDetails == nil
+	},
+		20*time.Minute, time.Second,
```

Comment:
the reason for such a long wait is that we need to wait for Fleet to start another round of checkin to recalculate state.

Reply:
also i need to add check for upgraded at/upgrade started fields but for that i need change in elastic-agent-libs. so followup in case i don't make the change in libs in time

Reply:
Yeah that is a long time, any ideas on how to shorten it? Reduce the check-in window?

Reply:
We had the same problem in https://github.com/elastic/elastic-agent/pull/7190 where all the timeouts got bumped to 5 minutes.

Shortening the checkin window works and I suspect is the fastest path. I like the idea of being able to trigger a checkin to happen and return on demand but that's a bit more work. Any action triggers the long poll to return so stuff like changing the log level triggers this in a hacky way, maybe a NOOP action would be the tidiest way to get an on-demand checkin response without changing too much.

Reply:
Also, we should evaluate if we can test this fix without an integration test. We already have really good upgrade integration test coverage so adding more isn't strictly necessary.

Would calling Upgrade() with a mock acker implementation be enough to validate this instead?

Reply:
remove Integ test and added unit test for coordinator and upgrade handler